compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -1938,10 +1938,43 @@ public void visit(BLangRecordDestructure recordDeStmt) {
             return;
         }
         typeChecker.checkExpr(recordDeStmt.expr, this.env);
+        validateBindingPatternForNonRequiredFields(recordDeStmt.varRef.recordRefFields, recordDeStmt.expr.getBType(),
+                recordDeStmt.expr.pos);
         checkRecordVarRefEquivalency(recordDeStmt.pos, recordDeStmt.varRef, recordDeStmt.expr.getBType(),
                                      recordDeStmt.expr.pos);
     }
 
+    private void validateBindingPatternForNonRequiredFields(List<BLangRecordVarRefKeyValue> fields,
+                                                            BType type, Location pos) {
+        type = types.getReferredType(type);
+
+        if (type.tag == TypeTags.MAP) {
+            for (BLangRecordVarRefKeyValue lhsField : fields) {
+                dlog.error(lhsField.variableName.pos,
+                        DiagnosticErrorCode.INVALID_FIELD_BINDING_PATTERN_WITH_NON_REQUIRED_FIELD);
+            }
+            return;
+        }
+        if (type.tag != TypeTags.RECORD) {
+            return;
+        }
+        BRecordType recordType = (BRecordType) type;
+        for (BLangRecordVarRefKeyValue lhsField : fields) {
+            if (recordType.fields.containsKey(lhsField.variableName.value)) {
+                if (Symbols.isOptional(recordType.fields.get(lhsField.variableName.value).symbol)) {
+                    dlog.error(lhsField.variableName.pos,
+                            DiagnosticErrorCode.INVALID_FIELD_BINDING_PATTERN_WITH_NON_REQUIRED_FIELD);
+                }
+            }
+            if (lhsField.variableReference.getKind() == NodeKind.RECORD_VARIABLE_REF) {
```

Comment:
Do we still need this?

Reply:
Yeah we need this to handle recordDestructure expressions.  An example like below
```
type Employee record {
    int a;
    record {
        int b?;
    } c;
};

public function main() {
    Employee emp = {a:4, c:{b:5}};
    int aa;
    int bb;
    {a:aa, c:{b:bb}} = emp;
}
```

Reply:
But the following then doesn't result in an error?

```ballerina
type Employee record {
    int a;
    record {
        int b?;
    }[1] c;
};

public function main() {
    Employee emp = {a: 4, c: [{b: 5}]};
    int aa;
    int bb;
    {a: aa, c: [{b: bb}]} = emp;
}
```

Reply:
Fixed it. Remove the method and move it to `checkRecordVarRefEquivalency` where not available fields are being checked.